l1/ethereum.go

```diff
@@ -0,0 +1,210 @@
+func (e *Ethereum) StarknetHead(ctx context.Context, l1BlockHash string) (*core.L1Head, error) {
+	var blockNumberString, blockHashString, stateRootString string
+
+	batch := []rpc.BatchElem{
+		{
+			Method: "eth_call",
+			Args: []any{
+				map[string]string{
+					"to": e.coreContractAddress.String(),
+					// EVM function selectors are the base64-encoded first four bytes of the
+					// Keccak-256 hash of the function signature (e.g., add(uint256,uint256)).
+					// https://docs.soliditylang.org/en/v0.8.10/abi-spec.html#function-selector
+					"data": "0x35befa5d", // Selector for stateBlockNumber()
+				},
+				map[string]string{
+					"blockHash": l1BlockHash,
+				},
+			},
+			Result: &blockNumberString,
+		},
+		{
+			Method: "eth_call",
+			Args: []any{
+				map[string]any{
+					"to":   e.coreContractAddress.String(),
+					"data": "0x382d83e3", // Selector for stateBlockHash().
+				},
+				map[string]string{
+					"blockHash": l1BlockHash,
+				},
+			},
+			Result: &blockHashString,
+		},
+		{
+			Method: "eth_call",
+			Args: []any{
+				map[string]string{
+					"to":   e.coreContractAddress.String(),
+					"data": "0x9588eca2", // Selector for stateRoot().
+				},
+				map[string]string{
+					"blockHash": l1BlockHash,
+				},
+			},
+			Result: &stateRootString,
+		},
+	}
+	if err := e.batchCallContext(ctx, batch); err != nil {
+		return nil, err
+	}
+
+	// We use slice[2:] to remove the `0x` prefix.
+
+	blockNumber, err := strconv.ParseUint(blockNumberString[2:], felt.Base16, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parse blockNumber %s: %w", blockNumberString, err)
+	}
+
+	blockHash, err := new(felt.Felt).SetString(blockHashString)
+	if err != nil {
+		return nil, fmt.Errorf("parse blockHash %s: %w", blockHashString, err)
+	}
+	stateRoot, err := new(felt.Felt).SetString(stateRootString)
+	if err != nil {
+		return nil, fmt.Errorf("parse stateRoot %s: %w", stateRootString, err)
+	}
+
+	return &core.L1Head{
+		BlockNumber: blockNumber,
+		BlockHash:   blockHash,
+		StateRoot:   stateRoot,
+	}, nil
+}
+
+// batchCallContext sends batch as an rpc request. If an error occurs while making the request,
+// it is returned. After receiving the result, it iterates over the batch and returns the first
+// error it finds. If no errors are found, it returns nil.
+func (e *Ethereum) batchCallContext(ctx context.Context, batch []rpc.BatchElem) error {
+	if err := e.client.BatchCallContext(ctx, batch); err != nil {
+		return fmt.Errorf("send batch request: %w", err)
+	}
+	for _, elem := range batch {
+		if elem.Error != nil {
+			return fmt.Errorf("batch response contains error: %w", elem.Error)
+		}
+	}
+	return nil
+}
+
+// ChainID implements part of the L1 interface.
+func (e *Ethereum) ChainID(ctx context.Context) (*big.Int, error) {
+	var result string
+	err := e.client.CallContext(ctx, &result, "eth_chainId")
+	if err != nil {
+		return nil, err
+	}
+	resultNoPrefix := result[2:] // Remove `0x` prefix.
```

Comment:
same for `TrimPrefix(...)` it's more explicit and has same number of allocations/copying